### PR TITLE
[Runtime] Fix Clang builds

### PR DESCRIPTION
**Pull request by dime10**
Fix some compilation warnings in the runtime provided by Clang which are now errors.

Also build the runtime with Clang in the regular PR check.

### CODE REVIEW

**Review by maliasadi — APPROVED** [review]
:tada:

File: .github/workflows/check-catalyst.yaml [hunk 1]
```diff
@@ -290,10 +290,12 @@ jobs:
...
-                            -DCMAKE_BUILD_TYPE=Release \
-                            -DLIGHTNING_GIT_TAG=$LIGHTNING_GIT_TAG_VALUE \
-                            -DQIR_STDLIB_LIB="$(pwd)/runtime/qir-stdlib-build/" \
-                            -DQIR_STDLIB_INCLUDES="$(pwd)/runtime/qir-stdlib-build/include/"
+                -DCMAKE_BUILD_TYPE=Release \
+                -DCMAKE_C_COMPILER=clang \
+                -DCMAKE_CXX_COMPILER=clang++ \
+                -DLIGHTNING_GIT_TAG=$LIGHTNING_GIT_TAG_VALUE \
+                -DQIR_STDLIB_LIB="$(pwd)/runtime/build-qir-stdlib/" \
+                -DQIR_STDLIB_INCLUDES="$(pwd)/runtime/build-qir-stdlib/include/"
```
Comment (maliasadi) [inline .github/workflows/check-catalyst.yaml:298]:
```suggestion
                -DQIR_STDLIB_LIB="$(pwd)/runtime/qir-stdlib-build/" \
                -DQIR_STDLIB_INCLUDES="$(pwd)/runtime/qir-stdlib-build/include/"
```

File: .github/workflows/check-catalyst.yaml [hunk 2]
```diff
@@ -341,22 +343,23 @@ jobs:
...
         if: ${{ matrix.simulator == 'lightning-kokkos' }}
         run: |
-            cmake -S runtime -B build -DCMAKE_BUILD_TYPE=Release \
-                                      -DENABLE_LIGHTNING_KOKKOS=ON \
-                                      -DQIR_STDLIB_LIB="$(pwd)/runtime/qir-stdlib-build/" \
-                                      -DQIR_STDLIB_INCLUDES="$(pwd)/runtime/qir-stdlib-build/include/" \
-                                      -G Ninja
+            cmake -S runtime -B build -G Ninja \
+                  -DCMAKE_BUILD_TYPE=Release \
+                  -DENABLE_LIGHTNING_KOKKOS=ON \
```
Comment (dime10) [inline .github/workflows/check-catalyst.yaml:360]: @maliasadi Shouldn't we include the git tag (`-DLIGHTNING_GIT_TAG=$LIGHTNING_GIT_TAG_VALUE \`) here for kokkos as well?

Comment (maliasadi) [reply .github/workflows/check-catalyst.yaml:360]: The Kokkos support needs to be updated/extended before using the git tag. We can address both when we revisit this support.

**Pull request merged**